### PR TITLE
Add code walkthrough to design section

### PR DESCRIPTION
**Pull request by sam-mccall**

### CODE REVIEW

**Comment by sam-mccall** [comment]
One thing I'm not sure of - should every mention of an identifier be a link? At the moment I try to introduce each one once and only link it there...

**Comment by ad1992** [comment]
@sam-mccall this looks great and really helpful. Thanks a lot for doing this 💯.

File: design/code.md [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+clangd shares code with other tools derived from the Clang compiler, these
+libraries live outside clangd.
+
+ - [syntax::TokenBuffer] captures token-boundary and preprocessing information
+   that clang itself doesn't preserve.
+ - [clang/Index/] implements an indexing-oriented traversal of Clang ASTs, which
+   is used (for now) in clangd's index.
+ - [tooling::CompilationDatabase] is the foundation for clangd's integration
+   with build systems.
+
```
Comment (kadircet) [inline design/code.md:218]: formatting libs ?

File: design/code.md [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+### Clang-tools libraries
+
+clangd shares code with other tools derived from the Clang compiler, these
+libraries live outside clangd.
+
+ - [syntax::TokenBuffer] captures token-boundary and preprocessing information
+   that clang itself doesn't preserve.
+ - [clang/Index/] implements an indexing-oriented traversal of Clang ASTs, which
+   is used (for now) in clangd's index.
```
Comment (kadircet) [inline design/code.md:215]: i don't think there's much value in mentioning future plans in there

File: design/code.md [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+allow emitting diagnosting and indexing ASTs.
+
+`ClangdServer` is fairly mechanical for the most part. The features are
+implemented in various other files, and the scheduling and AST building is
+done by `TUScheduler`, so largely it just binds these together.
+`TUScheduler` doesn't know about particular features (except diagnostics).
+
+### Compile commands
+
+Like other clang-based tools, clangd's uses clang's command-line syntax as its
```
Comment (sureyeaah) [inline design/code.md:69]: /s/clangd's/clangd

File: design/code.md [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+It mostly starts at the outside and works it way inward.
+
+The clangd code lives in the [llvm-project] repository under
+[clang-tools-extra/clangd][clangd]. We'll also mention some dependencies in
+other parts of llvm-project (such as clang).
+
+Links below point to the [woboq] code browser. It has nice cross-referenced navigation, but may show
+a slightly outdated version of the code.
+
+{% include toc.md %}
```
[the remark this answers is not in the document]
Comment (sam-mccall) [reply design/code.md:15]: `{% include ... %}` is a feature of jekyll. The file does exist.

File: design/code.md [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+hover information from this AST at offset 25"). ClangdServer exposes them as
+stateful, asynchronous functions ("compute hover information for the latest
+version of Foo.cpp at offset 25, call back when done") which is the LSP model.
+
+[TUScheduler] is responsible for keeping track of the latest version of each
+file, building and caching ASTs and preambles as inputs, and providing threads
+to run requests on in an appropriate sequence. (More details in
+[threads and request handling]).
+It also pushes certain events to `ClangdServer` via [ParsingCallbacks], to
+allow emitting diagnosting and indexing ASTs.
```
[the remark this answers is not in the document]
Comment (sam-mccall) [reply design/code.md:60]: Done. There is a bit under Index describing SymbolCollector, and also how FileIndex is built/used. Maybe it's too short, but I'm not sure what else to add - think it's best to leave this for a followup.

File: design/code.md [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+[CodeComplete.h] is mostly concerned with running clang in this mode,
+combining clang's results with index-based results, applying ranking, and
+converting to LSP's data model.
+
+The ranking is mostly implemented in [Quality.h], and the name-matching
+is done by [FuzzyMatcher].
+
+### Code actions
+
+Most code actions provided by `Tweak`s. These are small plugins that implement
```
[the remark this answers is not in the document]
Comment (sam-mccall) [reply design/code.md:124]: Done. (It is a link on the next line)

**Pull request merged**